src/odemis/acq/feature.py

```diff
@@ -119,3 +145,412 @@ def load_project_data(path: str) -> dict:
+
+        try:
+            # run the autofocus at the current position
+            logging.debug(f"running autofocus at {site.name.value} at focus: {self.focus.position.value}, req conf: {self.autofocus_conf_level}")
+            rel_rng = SAFE_REL_RANGE_DEFAULT
+            focus_rng = (self.focus.position.value["z"] + rel_rng[0], self.focus.position.value["z"] + rel_rng[1])
+            self._future._running_subf = AutoFocus(detector=self.streams[0].detector,
+                                                        emt=None,
+                                                        focus=self.focus,
+                                                        rng_focus=focus_rng)
+
+            # note: the auto focus moves the objective to the best position
+            foc_pos, foc_lev, conf = self._future._running_subf.result(timeout=900)
+            if conf >= self.autofocus_conf_level:
+
+                # update the feature focus position
+                pos = site.pos.value
+                site.pos.value = (pos[0], pos[1], foc_pos)  # NOTE: tuples cant do assignment so we need to replace the whole tuple
+                logging.debug(f"auto focus succeeded at {site.name.value} with conf:{conf}. new focus position: {foc_pos}")
+            else:
+                # if the confidence is low, restore the previous focus position
+                self._move_focus(site, {"z": site.pos.value[2]})
+                logging.debug(f"auto focus failed due at {site.name.value} with conf:{conf}. restoring focus position {site.pos.value[2]}")
+
+        except TimeoutError as e:
+            logging.debug(f"Timed out during autofocus at {site.name.value}. {e}")
+            self._future._running_subf.cancel()
+
+            # restore the previous focus position
+            self._move_focus(site, {"z": site.pos.value[2]})
+            logging.warning(f"auto focus timed out at {site.name.value}. restoring focus position {site.pos.value[2]}")
+
+    def _move_to_site(self, site: CryoFeature):
+        """
+        Move the stage to the given site and move the objective lens to position.
+        :param site: The site to move to.
+        :raises MoveError: if the stage failed to move to the given site.
+        """
+        # NOTE: site.pos is a tuple of (stage_x, stage_y, objective_z) coordinates
+        stage_pos = {
+            "x": site.pos.value[0],
+            "y": site.pos.value[1],
+        }
+        focus_pos = {"z": site.pos.value[2]}
+        logging.debug(f"For feature {site.name.value} moving the stage to {stage_pos} m")
+        self._future.running_subf = self.stage.moveAbs(stage_pos)
+
+        # estimate the time to move the stage
+        current_pos = (self.stage.position.value["x"], self.stage.position.value["y"])
+        site_pos = (stage_pos["x"], stage_pos["y"])
+        t = estimate_stage_movement_time(current_pos, site_pos)
+        t = t * 5 + 3 # adding extra margin
```

Comment:
I get the +3 for extra margin, but the *5 seems really excessive. Is the estimation function so far off?
Maybe a comment could explain the process better.

Reply:
The estimates are super unreliable because there are timing issues with the stage movements / position that we had to artificially wait the stage for a while when moving. There is sometimes a lot of artificial delay, that isn't predictable, so the *5 is a 'something has likely gone very wrong' timeout. I've added it as a constant with comment further up.